File: sdk/logging/dotnet/PrimusSaaS.Logging/Targets/ApplicationInsightsTarget.cs
using Microsoft.ApplicationInsights;
using Microsoft.ApplicationInsights.DataContracts;
using Microsoft.ApplicationInsights.Extensibility;
using PrimusSaaS.Logging.Core;

namespace PrimusSaaS.Logging.Targets;

/// <summary>
/// Target for sending logs to Azure Application Insights
/// </summary>
public class ApplicationInsightsTarget : ITarget
{
    private readonly TelemetryClient _telemetryClient;

    public ApplicationInsightsTarget(string? connectionString)
    {
        var config = TelemetryConfiguration.CreateDefault();
        
        if (!string.IsNullOrEmpty(connectionString))
        {
            config.ConnectionString = connectionString;
        }

        _telemetryClient = new TelemetryClient(config);
    }

    public void Write(LogEntry logEntry)
    {
        var severityLevel = MapSeverityLevel(logEntry.Level);
        var telemetry = new TraceTelemetry(logEntry.Message, severityLevel);

        // Add timestamp
        telemetry.Timestamp = logEntry.Timestamp;

        // Add context properties
        foreach (var kvp in logEntry.Context)
        {
            if (kvp.Value != null)
            {
                telemetry.Properties[kvp.Key] = kvp.Value.ToString();
            }
        }

        // Track exception if present in context
        if (logEntry.Context.TryGetValue("exception", out var exceptionObj) && exceptionObj is Exception ex)
        {
            var exceptionTelemetry = new ExceptionTelemetry(ex)
            {
                SeverityLevel = severityLevel,
                Timestamp = logEntry.Timestamp
            };

            // Copy properties to exception telemetry too
            foreach (var kvp in logEntry.Context)
            {
                if (kvp.Key != "exception" && kvp.Value != null)
                {
                    exceptionTelemetry.Properties[kvp.Key] = kvp.Value.ToString();
                }
            }

            _telemetryClient.TrackException(exceptionTelemetry);
        }
        else
        {
            _telemetryClient.TrackTrace(telemetry);
        }
    }

    public void Close()
    {
        _telemetryClient.Flush();
        // Allow some time for flushing
        Thread.Sleep(1000);
    }

    private SeverityLevel MapSeverityLevel(LogLevel level)
    {
        return level switch
        {
            LogLevel.Debug => SeverityLevel.Verbose,
            LogLevel.Info => SeverityLevel.Information,
            LogLevel.Warning => SeverityLevel.Warning,
            LogLevel.Error => SeverityLevel.Error,
            LogLevel.Critical => SeverityLevel.Critical,
            _ => SeverityLevel.Information
        };
    }
}

File: sdk/logging/nodejs/src/targets/ApplicationInsightsTarget.ts
import appInsights, { KnownSeverityLevel, TelemetryClient } from 'applicationinsights';
import { LogEntry } from '../core/LogEntry';
import { LogLevel } from '../core/LogLevel';
import { Target } from './Target';

export interface ApplicationInsightsTargetOptions {
  connectionString: string;
  roleName?: string;
}

/**
 * Sends structured logs to Azure Application Insights.
 */
export class ApplicationInsightsTarget implements Target {
  private client: TelemetryClient;

  constructor(options: ApplicationInsightsTargetOptions) {
    const client = appInsights.defaultClient ?? new TelemetryClient(options.connectionString);
    if (!appInsights.defaultClient) {
      appInsights
        .setup(options.connectionString)
        .setAutoCollectRequests(false)
        .setAutoCollectPerformance(false, false)
        .setAutoCollectExceptions(false)
        .setAutoCollectDependencies(false)
        .setAutoDependencyCorrelation(false)
        .setAutoCollectConsole(false)
        .setUseDiskRetryCaching(false)
        .start();
    }

    client.config.connectionString = options.connectionString;
    this.client = client;

    if (options.roleName) {
      this.client.context.tags[this.client.context.keys.cloudRole] = options.roleName;
    }
  }

  write(logEntry: LogEntry): void {
    try {
      this.client.trackTrace({
        message: logEntry.message,
        severity: mapSeverity(logEntry.level),
        properties: {
          ...logEntry.context,
          level: logEntry.level,
          timestamp: logEntry.timestamp
        }
      });
    } catch (err) {
      console.error('ApplicationInsightsTarget failed to send log', err);
    }
  }
}

function mapSeverity(level: LogLevel): KnownSeverityLevel {
  switch (level) {
    case LogLevel.DEBUG:
      return KnownSeverityLevel.Verbose;
    case LogLevel.INFO:
      return KnownSeverityLevel.Information;
    case LogLevel.WARNING:
      return KnownSeverityLevel.Warning;
    case LogLevel.ERROR:
      return KnownSeverityLevel.Error;
    case LogLevel.CRITICAL:
      return KnownSeverityLevel.Critical;
    default:
      return KnownSeverityLevel.Information;
  }
}
